File: js/samples/04.q.questBot/src/actions/mapAction.ts
import { TurnContext } from 'botbuilder';
import { ActionPlanner, Application } from '@microsoft/teams-ai';
import { ApplicationTurnState, IDataEntities, trimPromptResponse, updateDMResponse } from '../bot';
import * as responses from '../responses';

/**
 * @param {Application<ApplicationTurnState>} app The bot's application object.
 */
export function mapAction(app: Application<ApplicationTurnState>, planner: ActionPlanner<ApplicationTurnState>): void {
    app.ai.action('map', async (context: TurnContext, state: ApplicationTurnState, data: IDataEntities) => {
        const action = (data.operation ?? '').toLowerCase();
        switch (action) {
            case 'query':
                return await queryMap(app, context, state);
            default:
                await context.sendActivity(`[map.${action}]`);
                return '';
        }
    });
}

/**
 * @param {Application} app The bot's application object.
 * @param {TurnContext} context The current turn context.
 * @param {ApplicationTurnState} state The current application state.
 * @returns {Promise<boolean>} A boolean indicating whether the query was successful.
 */
async function queryMap(
    app: Application<ApplicationTurnState>,
    context: TurnContext,
    state: ApplicationTurnState
): Promise<string> {
    // Use the map to answer player
    // const newResponse = await app.ai.completePrompt(context, state, 'useMap');
    // if (newResponse) {
    //     await updateDMResponse(context, state, trimPromptResponse(newResponse).split('\n').join('<br>'));
    //     state.temp.playerAnswered = true;
    // } else {
    //     await updateDMResponse(context, state, responses.dataError());
    // }

    return '';
}

File: dotnet/samples/04.q.questBot/Actions/MapAction.cs
using Microsoft.Bot.Builder;
using Microsoft.TeamsAI.AI.Action;
using QuestBot.State;

namespace QuestBot.Actions
{
    public partial class QuestBotActions
    {
        [Action("map")]
        public async Task<bool> MapAsync([ActionTurnContext] ITurnContext turnContext, [ActionTurnState] QuestState turnState, [ActionEntities] Dictionary<string, object> entities)
        {
            ArgumentNullException.ThrowIfNull(turnContext);
            ArgumentNullException.ThrowIfNull(turnState);
            var data = GetData(entities);
            var action = (data.Operation ?? string.Empty).Trim().ToLowerInvariant();
            switch (action)
            {
                case "query":
                    // Use the map to answer player
                    var newResponse = await _application.AI.CompletePromptAsync(turnContext, turnState, "UseMap", null, default);
                    if (!string.IsNullOrEmpty(newResponse))
                    {
                        await UpdateDMResponseAsync(turnContext, turnState, string.Join("<br>", TrimPromptResponse(newResponse).Split('\n')));
                        turnState.Temp!.PlayerAnswered = true;
                    }
                    else
                    {
                        await UpdateDMResponseAsync(turnContext, turnState, ResponseGenerator.DataError());
                    }
                    return false;
                default:
                    await turnContext.SendActivityAsync($"[map.{action}]");
                    return true;
            }
        }
    }
}
